File: Cpp/081_SearchInRotatedSortedArray-II.cpp
class Solution{
public:
    bool search(vector<int>& nums, int target)
    {
        size_t len = nums.size();
        if(len == 0 || (len == 1 && nums[0] != target)) return false;
        if(nums[0] == target) return true;
        int front = 0, back = len - 1, middle = 0;
        while(front < back)
        {
            middle = (front + back) / 2;
            if(nums[middle] > nums[front]) front = middle;
            else if(nums[middle] == nums[front]) front++;
            else back = middle;
        }
        if(target >= nums[back] && target <= nums[len - 1]) return binarySearch(nums, back, len - 1, target);
        else if(target >= nums[0] && target <= nums[back - 1]) return binarySearch(nums, 0, back - 1, target);
        else return false;
    }

    bool binarySearch(vector<int>& nums, int front, int back, int target)
    {
        int i = front, j = back, middle = 0;
        while(i < j)
        {
            middle = (i + j) / 2;
            if(nums[middle] < target) i = middle + 1;
            else if(nums[middle] == target) return true;
            else j = middle;
        }
        return nums[j] == target;
    }
};

File: CSharp/081_SearchInRotatedSortedArray-II.cs
public class Solution {
    public bool Search(int[] nums, int target) {
        if(nums.Length == 0 || (nums.Length == 1 && nums[0] != target)) return false;
        if(nums[0] == target) return true;
        int front = 0, back = nums.Length - 1, middle = 0;
        while(front < back)
        {
            middle = (front + back) / 2;
            if(nums[middle] > nums[front]) front = middle;
            else if(nums[middle] == nums[front]) front++;
            else back = middle;
        }
        if(target >= nums[back] && target <= nums[nums.Length - 1])
        {
            if(Array.BinarySearch(nums, back, nums.Length - back, target) > 0) return true;
            else return false;
        }
        else if(target >= nums[0] && target <= nums[back - 1]) 
        {
            if(Array.BinarySearch(nums, 0, back, target) > 0) return true;
            else return false;
        }
        else return false;
    }
}